dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/patch/BulkPatchManagerTest.java

```diff
@@ -0,0 +1,257 @@
+        userD = createUser( "D" );
+        dataElementA = createDataElement( 'A' );
+        dataElementA
+            .setSharing( Sharing.builder().owner( userD.getUid() ).publicAccess( AccessStringHelper.DEFAULT ).build() );
+        dataElementA.setUid( "fbfJHSPpUQD" );
+        manager.save( dataElementA, false );
+
+        dataElementB = createDataElement( 'B' );
+        dataElementB.setUid( "cYeuwXTCPkU" );
+        dataElementB
+            .setSharing( Sharing.builder().owner( userD.getUid() ).publicAccess( AccessStringHelper.DEFAULT ).build() );
+        manager.save( dataElementB, false );
+
+        PeriodType periodType = periodService.getPeriodTypeByClass( MonthlyPeriodType.class );
+        dataSetA = createDataSet( 'A', periodType );
+        dataSetA.setUid( "em8Bg4LCr5k" );
+        dataSetA
+            .setSharing( Sharing.builder().owner( userD.getUid() ).publicAccess( AccessStringHelper.DEFAULT ).build() );
+        manager.save( dataSetA, false );
+    }
+
+    @Test
+    public void testApplyPatchOk()
+        throws IOException
+    {
+        final BulkJsonPatch bulkJsonPatch = jsonMapper.readValue(
+            new ClassPathResource( "patch/bulk_sharing_patch.json" ).getInputStream(),
+            BulkJsonPatch.class );
+
+        BulkPatchParameters patchParameters = BulkPatchParameters.builder()
+            .validators( BulkPatchValidators.sharingValidators() )
+            .build();
+
+        List<IdentifiableObject> patchedObjects = patchManager
+            .applyPatch( bulkJsonPatch, patchParameters );
+        assertEquals( 2, patchedObjects.size() );
+        assertTrue( aclService.canRead( userA, patchedObjects.get( 0 ) ) );
+        assertTrue( aclService.canRead( userA, patchedObjects.get( 1 ) ) );
+        assertFalse( aclService.canRead( userC, patchedObjects.get( 0 ) ) );
+    }
+
+    @Test
+    public void testApplyPatchInvalidClassName()
+        throws IOException
+    {
+        final BulkJsonPatch bulkJsonPatch = jsonMapper.readValue(
+            new ClassPathResource( "patch/bulk_sharing_patch_invalid_class_name.json" ).getInputStream(),
+            BulkJsonPatch.class );
+
+        BulkPatchParameters patchParameters = BulkPatchParameters.builder()
+            .validators( BulkPatchValidators.sharingValidators() )
+            .build();
+
+        List<IdentifiableObject> patchedObjects = patchManager
+            .applyPatch( bulkJsonPatch, patchParameters );
+        assertEquals( 0, patchedObjects.size() );
+        assertEquals( 1, patchParameters.getErrorReportsCount() );
+        assertEquals( 1, patchParameters.getErrorReportsCount( ErrorCode.E6002 ) );
+    }
+
+    @Test
+    public void testApplyPatchInvalidUid()
+        throws IOException
+    {
+        final BulkJsonPatch bulkJsonPatch = jsonMapper.readValue(
+            new ClassPathResource( "patch/bulk_sharing_patch_invalid_uid.json" ).getInputStream(),
+            BulkJsonPatch.class );
```

Comment:
Just noticed that these 3 lines are reoccurring in setup and could easily be made more readable with a `loadPatch(filename)` helper method.

Reply:
fixed